internal/cody/subscription.go

```diff
@@ -0,0 +1,147 @@
+package cody
+
+import (
+	"context"
+	"time"
+
+	"github.com/sourcegraph/sourcegraph/internal/database"
+	"github.com/sourcegraph/sourcegraph/internal/featureflag"
+	"github.com/sourcegraph/sourcegraph/internal/ssc"
+	"github.com/sourcegraph/sourcegraph/internal/types"
+	"github.com/sourcegraph/sourcegraph/lib/errors"
+)
+
+const USE_SSC_FEATURE_FLAG = "use-ssc-for-cody-subscription"
+
+type UserSubscriptionPlan string
+
+const (
+	UserSubscriptionPlanFree UserSubscriptionPlan = "free"
+	UserSubscriptionPlanPro  UserSubscriptionPlan = "pro"
+)
+
+type UserSubscription struct {
+	// Status is the current status of the subscription. "pending" means the user has no Cody Pro subscription.
+	// "pending" subscription will be removed post Feb 15, 2024. It is required to support users who have opted for
+	// Cody Pro Trial on dotcom.
+	Status ssc.SubscriptionStatus
+	// Plan is the plan the user is subscribed to. "free" or "pro".
+	Plan UserSubscriptionPlan
+	// ApplyProRateLimits is true if the user has a valid subscription status.
+	ApplyProRateLimits bool
+	// CurrentPeriodStartAt is the start date of the current billing cycle.
+	CurrentPeriodStartAt time.Time
+	// CurrentPeriodEndAt is the end date of the current billing cycle.
+	CurrentPeriodEndAt time.Time
+}
+
+// consolidateSubscriptionDetails consolidates the subscription details from dotcom and SSC.
+// Post Feb 15, 2024, this function will be removed and the logic will be simplified to us the SSC info only.
+func consolidateSubscriptionDetails(ctx context.Context, user types.User, subscription *ssc.Subscription) (*UserSubscription, error) {
+	now := currentTimeFromCtx(ctx)
+	febReleaseDate := time.Date(2024, 02, 14, 23, 59, 59, 59, now.Location())
```

Comment:
1. It is best to never, ever, ever deal with locals and timezones since it is an infinite source of bugs. All times are in UTC. Even if we wanted to be more friendly here, the behavior would be inconsistent and lead to bugs anyways.
2. "Midnight on 2/14" is a little too cute, IMHO. Let's just say 2/15.

```suggestion
	febReleaseDate := time.Date(2024, 02, 15, 0, 0, 0, 0, time.UTC)
```

Reply:
Actually, rather than hard-coding a date at all, why don't you introduce a new boolean feature flag. This way we can determine the behavior ourselves, and not be forced into it at a certain time.

e.g.

```go
offerCodyProFreeTrailFromDotcom := flags.Bool(true)`
```

Reply:
I'll update. 

But only that you don't need to submit any PRs to toggle feature flags. You should be "entitled" to do so from site-admin.

Reply:
@thenamankumar If we need to flip flags at certain times, please add the date/time for those actions in the [launch plan](https://docs.google.com/document/d/1bQkgy4g6wZyZOCgh5lGyjigiwIHNZyNH-Tc2vhXt-pU/edit#heading=h.v5j7go2zayy), too, with also marking who does it, or clearly marking that it's unassigned. (Let's assume that this PR will be merged.) Thanks!

Reply:
I will be resolving this as part of the next PR.

Reply:
I have decided the name of the feature flag to be `cody-pro-trial-ended` and won't be changing this as clients depends on it. 

Updated the doc here: https://docs.google.com/document/d/1p8V_c5wkmNgnCshRdlnjbHqxgkYLVlvpTyDdbHu4VSY/edit#heading=h.4r8kivvuokd0

Reply:
And not wasting any time, I am rather including the changes in this PR only.